{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input DNA = TTACGA\n",
      "\n",
      "Complement = AATGCT\n",
      "mRNA = AAUGCU\n",
      "Aminoacid = Asn - Ala\n"
     ]
    }
   ],
   "source": [
    "#DNA Translation\n",
    "complementary_pair = {\"A\" : \"T\", \n",
    "                      \"C\" : \"G\", \n",
    "                      \"G\" : \"C\", \n",
    "                      \"T\" : \"A\"}\n",
    "\n",
    "codon_table = {\n",
    "    'UUU': 'Phe', 'UUC': 'Phe', 'UUA': 'Leu', 'UUG': 'Leu',\n",
    "    'CUU': 'Leu', 'CUC': 'Leu', 'CUA': 'Leu', 'CUG': 'Leu',\n",
    "    'AUU': 'Ile', 'AUC': 'Ile', 'AUA': 'Ile', 'AUG': 'Met',\n",
    "    'GUU': 'Val', 'GUC': 'Val', 'GUA': 'Val', 'GUG': 'Val',\n",
    "    'UCU': 'Ser', 'UCC': 'Ser', 'UCA': 'Ser', 'UCG': 'Ser',\n",
    "    'CCU': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro',\n",
    "    'ACU': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr',\n",
    "    'GCU': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala',\n",
    "    'UAU': 'Tyr', 'UAC': 'Tyr', 'UAA': 'Stop', 'UAG': 'Stop',\n",
    "    'CAU': 'His', 'CAC': 'His', 'CAA': 'Gln', 'CAG': 'Gln',\n",
    "    'AAU': 'Asn', 'AAC': 'Asn', 'AAA': 'Lys', 'AAG': 'Lys',\n",
    "    'GAU': 'Asp', 'GAC': 'Asp', 'GAA': 'Glu', 'GAG': 'Glu',\n",
    "    'UGU': 'Cys', 'UGC': 'Cys', 'UGA': 'Stop', 'UGG': 'Trp',\n",
    "    'CGU': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg',\n",
    "    'AGU': 'Ser', 'AGC': 'Ser', 'AGA': 'Arg', 'AGG': 'Arg',\n",
    "    'GGU': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly'\n",
    "}\n",
    "\n",
    "def dna_complement(dna_seq: str):\n",
    "    complement = \"\"\n",
    "    hot_fix = dna_seq.upper()\n",
    "    for nucleotide in hot_fix:\n",
    "        if \"A\" == nucleotide:\n",
    "            complement += complementary_pair[\"A\"]\n",
    "        elif \"C\" == nucleotide:\n",
    "            complement += complementary_pair[\"C\"]\n",
    "        elif \"G\" == nucleotide:\n",
    "            complement += complementary_pair[\"G\"]\n",
    "        else:\n",
    "            complement += complementary_pair[\"T\"]\n",
    "    return complement\n",
    "\n",
    "\n",
    "def mrna_transformation(complement: str):\n",
    "    transformed = complement.replace(\"T\", \"U\")    \n",
    "    return transformed\n",
    "\n",
    "\n",
    "def amino_acid(mrna_seq):\n",
    "    if len(mrna_seq) % 3 != 0:\n",
    "        raise ValueError(\"mRNA sequence length must be divisible by 3\")\n",
    "    \n",
    "    codons = [mrna_seq[i:i+3] for i in range(0, len(mrna_seq), 3)]\n",
    "\n",
    "    amino_acids = [codon_table.get(codon, 'Unknown') for codon in codons]\n",
    "\n",
    "    joiner = \" - \" if len(amino_acids) == 2 else \" - \" if len(amino_acids) == 3 else \" - \"\n",
    "\n",
    "    return joiner.join(amino_acids)\n",
    "    \n",
    "\n",
    "#test case\n",
    "dna = input(\"enter dna sequence: \")\n",
    "\n",
    "comp = dna_complement(dna)\n",
    "mrna = mrna_transformation(comp)\n",
    "amino = amino_acid(mrna)\n",
    "\n",
    "print(f\"Input DNA = {dna}\\n\")\n",
    "print(f\"Complement = {comp}\")\n",
    "print(f\"mRNA = {mrna}\")\n",
    "print(f\"Aminoacid = {amino}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino Acid = WYW\n",
      "\n",
      "mRNA: UGGUAUUGG\n",
      "  UGG: 2\n",
      "  UAU: 1\n",
      "\n",
      "\n",
      "mRNA: UGGUACUGG\n",
      "  UGG: 2\n",
      "  UAC: 1\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#RNA Frequency\n",
    "codons = {\n",
    "    'F': ['UUU', 'UUC'], 'L': ['UUA', 'UUG', 'CUU', 'CUC', 'CUA', 'CUG'],\n",
    "    'I': ['AUU', 'AUC', 'AUA'], 'M': ['AUG'],\n",
    "    'V': ['GUU', 'GUC', 'GUA', 'GUG'], 'S': ['UCU', 'UCC', 'UCA', 'UCG', 'AGU', 'AGC'],\n",
    "    'P': ['CCU', 'CCC', 'CCA', 'CCG'], 'T': ['ACU', 'ACC', 'ACA', 'ACG'],\n",
    "    'A': ['GCU', 'GCC', 'GCA', 'GCG'], 'Y': ['UAU', 'UAC'],\n",
    "    'H': ['CAU', 'CAC'], 'Q': ['CAA', 'CAG'], 'N': ['AAU', 'AAC'],\n",
    "    'K': ['AAA', 'AAG'], 'D': ['GAU', 'GAC'], 'E': ['GAA', 'GAG'],\n",
    "    'C': ['UGU', 'UGC'], 'W': ['UGG'], 'R': ['CGU', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "    'G': ['GGU', 'GGC', 'GGA', 'GGG'], 'Stop': ['UAA', 'UAG', 'UGA']\n",
    "}\n",
    "\n",
    "\n",
    "def generate_combinations(amino_acid_seq):\n",
    "    if not amino_acid_seq:\n",
    "        return ['']  \n",
    "    \n",
    "    first_aa = amino_acid_seq[0]\n",
    "    possible_codons = codons.get(first_aa, [])\n",
    "    \n",
    "    remaining_combinations = generate_combinations(amino_acid_seq[1:])\n",
    "    \n",
    "    all_combinations = []\n",
    "    for codon in possible_codons:\n",
    "        for combination in remaining_combinations:\n",
    "            all_combinations.append(codon + combination)\n",
    "    \n",
    "    return all_combinations\n",
    "\n",
    "def count_codons(mrna_seq):\n",
    "    codon_count = {}\n",
    "    for i in range(0, len(mrna_seq), 3):\n",
    "        codon = mrna_seq[i:i+3]\n",
    "        codon_count[codon] = codon_count.get(codon, 0) + 1\n",
    "    return codon_count\n",
    "\n",
    "\n",
    "def rna_frequency(amino_acids: str):\n",
    "    all_mrna_combinations = generate_combinations(amino_acids)\n",
    "    for mrna_seq in all_mrna_combinations:\n",
    "        codon_count = count_codons(mrna_seq)\n",
    "        print(f\"mRNA: {mrna_seq}\")\n",
    "        for codon, count in codon_count.items():\n",
    "            print(f\"  {codon}: {count}\")\n",
    "\n",
    "input_test = input(\"Enter amino acid\")\n",
    "\n",
    "print(f\"Amino Acid = {input_test.upper()}\\n\")\n",
    "rna_frequency(input_test.upper())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
